activerecord/lib/active_record/persistence.rb

```diff
@@ -359,7 +359,12 @@ def destroy!
     # Any change to the attributes on either instance will affect both instances.
     # If you want to change the sti column as well, use #becomes! instead.
     def becomes(klass)
-      became = klass.new
+      if klass.descends_from_active_record? &&
+         klass.attribute_names.include?(klass.inheritance_column)
+        became = klass.new(klass.inheritance_column => klass.name)
+      else
+        became = klass.new
+      end
```

Comment:
How about avoiding `klass.new` simply?

```diff
diff --git a/activerecord/lib/active_record/persistence.rb b/activerecord/lib/active_record/persistence.rb
index 462e5e7aaf..e5fa4a925e 100644
--- a/activerecord/lib/active_record/persistence.rb
+++ b/activerecord/lib/active_record/persistence.rb
@@ -359,7 +359,8 @@ def destroy!
     # Any change to the attributes on either instance will affect both instances.
     # If you want to change the sti column as well, use #becomes! instead.
     def becomes(klass)
-      became = klass.new
+      became = klass.allocate
+      became.send(:initialize)
       became.instance_variable_set("@attributes", @attributes)
       became.instance_variable_set("@mutations_from_database", @mutations_from_database) if defined?(@mutations_from_database)
       became.instance_variable_set("@changed_attributes", attributes_changed_by_setter)
```

Reply:
I'm not familiar with the cons of `klass.new` and `.allocate`. Let me read more about it and I'll change the code.